### PR TITLE
Send test case events to data collector process.

### PR DESCRIPTION
**Pull request by harshjain2**
Send test case events to data collector process.

Related to : https://github.com/Microsoft/vstest/issues/309

### CODE REVIEW

File: src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionRequestHandler.cs [hunk 1]
```diff
@@ -58,9 +74,23 @@ internal DataCollectionRequestHandler(ICommunicationManager communicationManager
...
-        /// Gets singleton instance of DataCollectionRequestHandler.
+        /// Gets the singleton instance of DataCollectionRequestHandler.
+        /// </summary>
+        public static DataCollectionRequestHandler Instance { get; private set; }
+
+        /// <summary>
+        /// Creates singleton instance of DataCollectionRequestHandler.
         /// </summary>
-        public static DataCollectionRequestHandler CreateInstance(ICommunicationManager communicationManager, IMessageSink messageSink)
+        /// <param name="communicationManager">
```
Comment (codito) [inline src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionRequestHandler.cs]: nit: auto generated docs, please clean if required.

File: src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionRequestHandler.cs [hunk 2]
```diff
@@ -15,40 +16,55 @@ namespace Microsoft.VisualStudio.TestPlatform.CommunicationUtilities.DataCollect
...
     using Microsoft.VisualStudio.TestPlatform.ObjectModel.Logging;
-    using System.Threading;
 
     /// <summary>
     /// Handles test session events received from vstest console process.
     /// </summary>
     internal class DataCollectionRequestHandler : IDataCollectionRequestHandler, IDisposable
     {
+        private const int dataCollectionConnTimeout = 15 * 1000;
+        private static readonly object obj = new object();
```
Comment (codito) [inline src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionRequestHandler.cs]: Better to name as `syncObject` (as used elsewhere in code base).

File: src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionRequestHandler.cs [hunk 3]
```diff
@@ -115,26 +145,59 @@ public void ProcessRequests()
...
-                                 // todo : decide the time out for this wait as it has to wait till test execution process get created.
-                                 this.dataCollectionTestCaseEventHandler.WaitForRequestHandlerConnection(0);
-                                 this.dataCollectionTestCaseEventHandler.ProcessRequests();
-                             }, this.cancellationTokenSource.Token);
+                            this.testCaseEventMonitorTask = Task.Factory.StartNew(
+                                () =>
+                                    {
+                                        try
+                                        {
+                                            if (this.dataCollectionTestCaseEventHandler.WaitForRequestHandlerConnection(dataCollectionConnTimeout))
```
Comment (codito) [inline src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionRequestHandler.cs]: Do we support `VSTEST_COLLECTOR_DEBUG` for debugging data collectors? If yes, we need to add a note to user here that data collector process is waiting for debugger attach.

File: src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionRequestHandler.cs [hunk 4]
```diff
@@ -144,29 +207,44 @@ public void ProcessRequests()
...
                         catch (Exception ex)
                         {
-                            EqtTrace.Error("DataCollectionRequestHandler.ProcessRequests : {0}", ex.Message);
+                            if (EqtTrace.IsErrorEnabled)
+                            {
+                                EqtTrace.Error("DataCollectionRequestHandler.ProcessRequests : {0}", ex.Message);
+                            }
                         }
 
                         var attachments = this.dataCollectionManager.SessionEnded(isCancelled);
```
Comment (codito) [inline src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionRequestHandler.cs]: Suppose this is a heavy operation, the message loop will not process any other messages (including cancel) until this is complete.

File: src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionTestCaseEventHandler.cs [hunk 5]
```diff
@@ -55,15 +62,69 @@ public void Close()
...
-            this.communicationManager?.StopServer();
+            do
+            {
+                var message = this.communicationManager.ReceiveMessage();
+                switch (message.MessageType)
+                {
+                    case MessageType.BeforeTestCaseStart:
+                        if (EqtTrace.IsInfoEnabled)
+                        {
+                            EqtTrace.Info("DataCollectionTestCaseEventHandler : Test case starting.");
```
Comment (codito) [inline src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionTestCaseEventHandler.cs]: nit: good to check if we are using a space before `:` in log messages for consistency.

File: src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionTestCaseEventSender.cs [hunk 6]
```diff
@@ -12,24 +13,61 @@ namespace Microsoft.VisualStudio.TestPlatform.CommunicationUtilities
...
-        public DataCollectionTestCaseEventSender(ICommunicationManager communicationManager)
+        protected DataCollectionTestCaseEventSender(ICommunicationManager communicationManager)
         {
             this.communicationManager = communicationManager;
         }
 
+        /// <summary>
+        /// Gets the singleton instance of DataCollectionTestCaseEventSender.
+        /// </summary>
+        internal static DataCollectionTestCaseEventSender Instance { get; private set; }
```
Comment (codito) [inline src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionTestCaseEventSender.cs]: Why singleton?

File: src/Microsoft.TestPlatform.CrossPlatEngine/Execution/BaseRunTests.cs [hunk 7]
```diff
@@ -88,24 +90,27 @@ protected BaseRunTests(string runSettings, TestExecutionContext testExecutionCon
...
             this.inProcDataCollectionExtensionManager = new InProcDataCollectionExtensionManager(runSettings, testRunCache, this.dataCollectionTestCaseEventManager);
 
-
-
-            // Verify if datacollection is enabled and wrap the testcasehandler around to get the events 
-            if (inProcDataCollectionExtensionManager.IsInProcDataCollectionEnabled)
+            if (DataCollectionTestCaseEventSender.Instance != null)
             {
-                this.testCaseEventsHandler = new TestCaseEventsHandler(this.dataCollectionTestCaseEventManager, this.testCaseEventsHandler);
+                this.outOfProcDataCollectionManager = new ProxyOutOfProcDataCollectionManager(DataCollectionTestCaseEventSender.Instance, this.dataCollectionTestCaseEventManager);
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Execution/BaseRunTests.cs]: This code is not testable. `DataCollectionTestCaseEventSender` is hardcoded static.

File: src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionTestCaseEventSender.cs [hunk 8]
```diff
@@ -12,24 +13,61 @@ namespace Microsoft.VisualStudio.TestPlatform.CommunicationUtilities
...
+
+        /// <summary>
+        /// Gets singleton instance of DataCollectionRequestHandler.
+        /// </summary>
+        /// <param name="communicationManager">
+        /// The communication Manager.
+        /// </param>
+        public static DataCollectionTestCaseEventSender Create(ICommunicationManager communicationManager)
```
Comment (codito) [inline src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionTestCaseEventSender.cs]: Why is `ICommunicationManager` argument required? Don't see it used anywhere.

File: src/Microsoft.TestPlatform.CommunicationUtilities/Friends.cs [hunk 9]
```diff
@@ -8,6 +8,8 @@
...
 [assembly: InternalsVisibleTo("Microsoft.TestPlatform.CrossPlatEngine.UnitTests, PublicKey=002400000480000094000000060200000024000052534131000400000100010007d1fa57c4aed9f0a32e84aa0faefd0de9e8fd6aec8f87fb03766c834c99921eb23be79ad9d5dcc1dd9ad236132102900b723cf980957fc4e177108fc607774f29e8320e92ea05ece4e821c0a5efe8f1645c4c0c93c1ab99285d622caa652c1dfad63d745d6f2de5f17e5eaf0fc4963d261c8a12436518206dc093344d5ad293")]
 [assembly: InternalsVisibleTo("datacollector, PublicKey=002400000480000094000000060200000024000052534131000400000100010007d1fa57c4aed9f0a32e84aa0faefd0de9e8fd6aec8f87fb03766c834c99921eb23be79ad9d5dcc1dd9ad236132102900b723cf980957fc4e177108fc607774f29e8320e92ea05ece4e821c0a5efe8f1645c4c0c93c1ab99285d622caa652c1dfad63d745d6f2de5f17e5eaf0fc4963d261c8a12436518206dc093344d5ad293")]
+[assembly: InternalsVisibleTo("testhost.x86, PublicKey=002400000480000094000000060200000024000052534131000400000100010007d1fa57c4aed9f0a32e84aa0faefd0de9e8fd6aec8f87fb03766c834c99921eb23be79ad9d5dcc1dd9ad236132102900b723cf980957fc4e177108fc607774f29e8320e92ea05ece4e821c0a5efe8f1645c4c0c93c1ab99285d622caa652c1dfad63d745d6f2de5f17e5eaf0fc4963d261c8a12436518206dc093344d5ad293")]
```
Comment (codito) [inline src/Microsoft.TestPlatform.CommunicationUtilities/Friends.cs]: Why is this required?

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyExecutionManagerWithDataCollection.cs [hunk 10]
```diff
@@ -131,6 +133,17 @@ public override void Close()
         {
             base.Close();
         }
+
+        /// <inheritdoc />
+        protected override TestProcessStartInfo GetTestHostProcessStartInfo(
+            IEnumerable<string> sources,
+            TestRunnerConnectionInfo connectionInfo)
+        {
+            var testHostProcessStartInfo = this.testHostManager.GetTestHostProcessStartInfo(sources, this.dataCollectionParameters.EnvironmentVariables, connectionInfo);
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyExecutionManagerWithDataCollection.cs]: Now we're leaking the knowledge of `ITestHostManager` to every proxy. Is it a better abstraction to just allow derived classes to update the `TestProcessStartInfo`?

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs [hunk 11]
```diff
@@ -26,15 +26,13 @@ public abstract class ProxyOperationManager
     {
         private readonly ITestHostManager testHostManager;
 
-        private bool initialized;
+        protected bool initialized;
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs]: Why is this change required?

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollectionExtensionManager.cs [hunk 12]
```diff
@@ -61,7 +61,7 @@ public InProcDataCollectionExtensionManager(string runSettings, ITestRunCache te
             // Initialize InProcDataCollectors
             this.InitializeInProcDataCollectors(runSettings);
 
-            FlushResultTestResultPoperty = TestProperty.Register(id: "allowTestResultFlush", label: "allowTestResultFlush", category: string.Empty, description: string.Empty, valueType: typeof(bool), validateValueCallback: null, attributes: TestPropertyAttributes.None, owner: typeof(TestCase));
+            FlushResultTestResultPoperty = TestProperty.Register(id: "allowTestResultFlush", label: "allowTestResultFlush", category: string.Empty, description: string.Empty, valueType: typeof(bool?), validateValueCallback: null, attributes: TestPropertyAttributes.None, owner: typeof(TestCase));
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollectionExtensionManager.cs]: Why should the `valueType` be nullable?

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollectionExtensionManager.cs [hunk 13]
```diff
@@ -95,6 +94,31 @@ public void FlushLastChunkResults()
         }
 
         /// <summary>
+        /// The create data collector.
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollectionExtensionManager.cs]: nit: auto generated docs, please update as needed.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Execution/BaseRunTests.cs [hunk 14]
```diff
@@ -88,24 +90,27 @@ protected BaseRunTests(string runSettings, TestExecutionContext testExecutionCon
...
+            this.testRunCache = new TestRunCache(this.testExecutionContext.FrequencyOfRunStatsChangeEvent, testExecutionContext.RunStatsChangeEventTimeout, this.OnCacheHit);
             this.dataCollectionTestCaseEventManager = new DataCollectionTestCaseEventManager();
 
             this.inProcDataCollectionExtensionManager = new InProcDataCollectionExtensionManager(runSettings, testRunCache, this.dataCollectionTestCaseEventManager);
 
-
-
-            // Verify if datacollection is enabled and wrap the testcasehandler around to get the events 
-            if (inProcDataCollectionExtensionManager.IsInProcDataCollectionEnabled)
+            if (DataCollectionTestCaseEventSender.Instance != null)
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Execution/BaseRunTests.cs]: When will the `Instance` be null?

File: src/testhost.x86/DefaultEngineInvoker.cs [hunk 15]
```diff
@@ -32,6 +33,8 @@ internal class DefaultEngineInvoker :
 
         private const string LogFileArgument = "--diag";
 
+        private const string DataCollectionPortArgument = "--datacollectionport";
```
Comment (codito) [inline src/testhost.x86/DefaultEngineInvoker.cs]: Not sure if it is a good idea to reuse testhost program logic for data collectors. Test host has complex sequences including app domain etc., datacollector doesn't need them.

File: test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/DataCollectionRequestHandlerTests.cs [hunk 16]
```diff
@@ -48,22 +49,22 @@ public void CreateInstanceShouldThrowExceptinIfInstanceMessageSinkIsNull()
...
+            var result = DataCollectionRequestHandler.Create(this.mockCommunicationManager.Object, this.mockMessageSink.Object);
 
             Assert.AreEqual(result, DataCollectionRequestHandler.Instance);
         }
 
         [TestMethod]
         public void InitializeCommunicationShouldInitializeCommunication()
         {
-            var requestHandler = new DataCollectionRequestHandler(this.mockCommunicationManager.Object, this.mockMessageSink.Object, this.mockDataCollectionManager.Object, this.mockDataCollectionTestCaseEventHandler.Object);
+            var requestHandler = new TestableDataCollectionRequestHandler(this.mockCommunicationManager.Object, this.mockMessageSink.Object, this.mockDataCollectionManager.Object, this.mockDataCollectionTestCaseEventHandler.Object);
```
Comment (codito) [inline test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/DataCollectionRequestHandlerTests.cs]: Could be moved to an instance variable?

File: test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/DataCollectionTestCaseEventHandlerTests.cs [hunk 17]
```diff
@@ -4,28 +4,38 @@
...
             this.mockCommunicationManager = new Mock<ICommunicationManager>();
+            this.mockDataCollectionManager = new Mock<IDataCollectionManager>();
         }
 
         [TestMethod]
         public void InitializeShouldInitializeConnection()
         {
             this.mockCommunicationManager.Setup(x => x.HostServer()).Returns(1);
-            var requestHandler = new DataCollectionTestCaseEventHandler(this.mockCommunicationManager.Object);
+            var requestHandler = new DataCollectionTestCaseEventHandler(this.mockCommunicationManager.Object, new Mock<IDataCollectionManager>().Object);
```
Comment (codito) [inline test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/DataCollectionTestCaseEventHandlerTests.cs]: Could be moved as an instance variable?

File: test/datacollector.UnitTests/DataCollectionManagerTests.cs [hunk 18]
```diff
@@ -348,6 +348,111 @@ public void SessionEndedShouldCancelProcessingAttachmentRequestsIfSessionIsCance
...
         }
+
+        #region TestCaseEventsTest
+
+        [TestMethod]
+        public void TestCaseStartedShouldSendEventToDataCollector()
+        {
+            var isStartInvoked = false;
+            var runSettings = string.Format(this.defaultRunSettings, this.dataCollectorSettings);
+            this.mockDataCollector.Setup(
```
Comment (codito) [inline test/datacollector.UnitTests/DataCollectionManagerTests.cs]: DRY

**Review by codito — CHANGES_REQUESTED** [review]
Suggested few changes, need more clarification in some cases.

File: src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionRequestHandler.cs (hunk 2 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionRequestHandler.cs]: done.

File: src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionRequestHandler.cs (hunk 3 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionRequestHandler.cs]: Added a todo for now in Program.cs for datacollector.exe. This will be implemented in upcoming PR.

File: src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionTestCaseEventSender.cs (hunk 6 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionTestCaseEventSender.cs]: DataCollectionTestCaseEventSender is used to send events to DataCollector Process from TestExecution Process. We get DataCollector Port as arguments which is where we are instantiating object of this class and establish the socket communication. Later on, this class is used to send the events, which starts when the TestExecution is in progress. Therefore, this class is exposed as singleton.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyExecutionManagerWithDataCollection.cs (hunk 10 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyExecutionManagerWithDataCollection.cs]: done.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Execution/BaseRunTests.cs (hunk 14 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Execution/BaseRunTests.cs]: When DataCollection Port is not provided, the instance will be null.

File: test/datacollector.UnitTests/DataCollectionManagerTests.cs (hunk 18 above)
Comment (harshjain2) [inline test/datacollector.UnitTests/DataCollectionManagerTests.cs]: Nope.

File: test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/DataCollectionTestCaseEventHandlerTests.cs (hunk 17 above)
Comment (harshjain2) [inline test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/DataCollectionTestCaseEventHandlerTests.cs]: Done.

File: test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/DataCollectionRequestHandlerTests.cs (hunk 16 above)
Comment (harshjain2) [inline test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/DataCollectionRequestHandlerTests.cs]: Done.

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollectionExtensionManager.cs (hunk 13 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollectionExtensionManager.cs]: Done.

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollectionExtensionManager.cs (hunk 12 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollectionExtensionManager.cs]: Value of FlushResultTestResultPoperty is null when inproc data collection is not enabled.
Otherwise, we might always get false by default, which will disable the flushing of test results permanently, vstest.console process will never get the test results even though tests are executed as expected.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs (hunk 11 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs]: not required.

File: src/Microsoft.TestPlatform.CommunicationUtilities/Friends.cs (hunk 9 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CommunicationUtilities/Friends.cs]: This is required for exposing **DataCollectionTestCaseEventSender** to testhostprocess.

File: src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionTestCaseEventHandler.cs (hunk 5 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionTestCaseEventHandler.cs]: Done.

File: src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionRequestHandler.cs (hunk 1 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionRequestHandler.cs]: Done.

File: src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionRequestHandler.cs (hunk 4 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionRequestHandler.cs]: Created a task so that this can be a non-blocking call.

**Comment by harshjain2** [comment]
@dotnet-bot test this please

File: src/Microsoft.TestPlatform.CrossPlatEngine/Execution/BaseRunTests.cs (hunk 7 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Execution/BaseRunTests.cs]: We will take this up in next PR.

File: src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionRequestHandler.cs [hunk 19]
```diff
@@ -218,9 +219,16 @@ public void ProcessRequests()
...
-                        var attachments = this.dataCollectionManager.SessionEnded(isCancelled);
+                        var attachmentsTask = Task<Collection<AttachmentSet>>.Factory.StartNew(
+                            () =>
+                                {
+                                    return this.dataCollectionManager.SessionEnded(isCancelled);
+                                },
+                            this.cancellationTokenSource.Token);
 
-                        this.communicationManager.SendMessage(MessageType.AfterTestRunEndResult, attachments);
+                        attachmentsTask.Wait(this.cancellationTokenSource.Token);
```
Comment (codito) [inline src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionRequestHandler.cs]: This will cause two threads to wait?

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs [hunk 20]
```diff
@@ -165,23 +167,18 @@ public virtual void Close()
         #endregion
 
         /// <summary>
-        /// The get test host process start info.
+        /// This method is exposed to enable drived classes to modify TestProcessStartInfo. E.g. DataCollection need additional environment variables to be passed, etc.
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs]: nit: spell

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollectionExtensionManager.cs [hunk 21]
```diff
@@ -61,7 +61,7 @@ public InProcDataCollectionExtensionManager(string runSettings, ITestRunCache te
             // Initialize InProcDataCollectors
             this.InitializeInProcDataCollectors(runSettings);
 
-            FlushResultTestResultPoperty = TestProperty.Register(id: "allowTestResultFlush", label: "allowTestResultFlush", category: string.Empty, description: string.Empty, valueType: typeof(bool?), validateValueCallback: null, attributes: TestPropertyAttributes.None, owner: typeof(TestCase));
+            FlushResultTestResultPoperty = TestProperty.Register(id: "allowTestResultFlush", label: "allowTestResultFlush", category: string.Empty, description: string.Empty, valueType: typeof(bool), validateValueCallback: null, attributes: TestPropertyAttributes.None, owner: typeof(TestCase));
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollectionExtensionManager.cs]: Please add a comment for the need of this property and expected behavior.

File: src/Microsoft.TestPlatform.CommunicationUtilities/Friends.cs (hunk 9 above)
Comment (codito) [inline src/Microsoft.TestPlatform.CommunicationUtilities/Friends.cs]: :) Since `CommunicationUtilities` is not a public library, isn't making types `internal` and adding `InternalsVisibleTo` similar to making it `public`?

File: src/Microsoft.TestPlatform.CrossPlatEngine/Execution/BaseRunTests.cs (hunk 14 above)
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Execution/BaseRunTests.cs]: I believe this logic should move up to ExecutionManager.

File: test/datacollector.UnitTests/DataCollectionManagerTests.cs (hunk 18 above)
Comment (codito) [inline test/datacollector.UnitTests/DataCollectionManagerTests.cs]:
```
SetupMockDataCollector(Func callback)

// Let the callback do whatever event needs to be subscribed.
```

Comment (codito) [inline test/datacollector.UnitTests/DataCollectionManagerTests.cs]: `runSettings` is using instance variables, why can't it be initialized in ctor?

**Review by codito — APPROVED** [review]
Approved with few suggestions.

File: src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionTestCaseEventSender.cs (hunk 8 above)
Comment (codito) [inline src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionTestCaseEventSender.cs]: Please clean this before merge.

File: src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionTestCaseEventSender.cs (hunk 6 above)
Comment (codito) [inline src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionTestCaseEventSender.cs]: Why can't this be passed along with DI?

Comment (codito) [inline src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionTestCaseEventSender.cs]: OK to take up when we make the BaseRunTests testable in next PR.

Comment (harshjain2) [inline src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionTestCaseEventSender.cs]: Added a todo here. Will take that up with BaseRunTests refactoring.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Execution/BaseRunTests.cs (hunk 14 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Execution/BaseRunTests.cs]: Will take this up with BaseRunTests refactoring.

File: src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionTestCaseEventSender.cs (hunk 8 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionTestCaseEventSender.cs]: Done.

File: src/Microsoft.TestPlatform.CommunicationUtilities/Friends.cs (hunk 9 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CommunicationUtilities/Friends.cs]: Done.

File: test/datacollector.UnitTests/DataCollectionManagerTests.cs (hunk 18 above)
Comment (harshjain2) [inline test/datacollector.UnitTests/DataCollectionManagerTests.cs]: Done.
